### PR TITLE
Keccak Challenge-API integration

### PR DESCRIPTION
**Pull request by CPerezz**
This introduces the challenge API inclusion into the all three different Keccak circuit implementations. Resolves: #751 
This introduces 2 different challenges. One for the inputs and another for the table-exposed value.

The changes consists on:
1- Swap the meta.advice_column for meta.advice_column_in(SecondPhase).
2- Replace the randomnes, which is hardcoded now in the circuit by the Challenges.
Including it in the configure and witness_generation functions.
3- Test with a custom Challenges which has a fixed value.

From this work, two main issues/tech-debt derive.
1.  >I made an effort to migrate `Part<F>` and `PartValue<F>` to use the
`Value<F>` halo2 API. But since the struct doesn't implement `PartialEq`
it's challenging to do it. As I do need to operate with it quite a lot.
>
>Also, there's another caveat which is that it feels really tricky to
extract the byte representation from a `Value<F>`. It feels like if
you're cheating to the API constrantly (using `.map()` to gain access to
the underlying F so that you can do extra things with it such as
decompose it into bytes).
>
>For these reasons, right now `extract_field` was added. An issue will be
filled to track the tech-debt left here and pay it once we maybe we add
the required functionalities to `halo2::Value<F>`.

2. The `SuperCircuit` is using now the slowest and worst performant of the 3 keccak implementations. So I ported the three so that we can run benchmarks for them with the SuperCircuit which will help to decide which one we keep and which ones we remove to reduce code manteinance.
After a discussion with @Brechtpd on this topic, we both agreed that `multi-packed` approach should be the one to keep. But I think is worth a benchmark before we take that path removing the two other implementations from the repo.

Tech-debt issues generated from this PR:
- #941
- #942

### CODE REVIEW

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs [hunk 1]
```diff
@@ -178,6 +181,24 @@ impl<F: FieldExt> Cell<F> {
...
+    pub(crate) fn assign_value(&self, region: &mut KeccakRegion<F>, offset: i32, value: Value<F>) {
+        // This is really ugly. But since there's no way to easily adapt the CellManager
+        // API customized for this impl specifically, for now I'm opening the
+        // value and extracting it. Once https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/933 is resolved,
+        // this shouldn't be needed.
+        let mut value_f = F::zero();
+        value.map(|f| {
+            value_f = f;
+            f
+        });
```
Comment (han0110) [inline zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:194]: Perhaps we could reuse `extract_field`?

File: zkevm-circuits/src/util.rs [hunk 2]
```diff
@@ -54,6 +54,7 @@ pub struct Challenges<T = Challenge> {
 impl Challenges {
     /// Construct `Challenges` by allocating challenges in specific phases.
     pub fn construct<F: FieldExt>(meta: &mut ConstraintSystem<F>) -> Self {
+        let _dummy_col = meta.advice_column();
```
Comment (han0110) [inline zkevm-circuits/src/util.rs:57]: I think this still adds small overhead for both prover and verifier, it'd be nice if we could avoid this workaround (we just need to call `{KeccakBitConfig,KeccakPackedConfig}::configure(meta, challenge_exprs)` before `Challenges::construct(meta)`).

But I understand it's annoying for people to encounter the panic, so I'm also fine to keep it here.

**Review by han0110 — APPROVED** [review]
LGTM! Nice work! Only left some minor comments.

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs (hunk 1 above)
Comment (CPerezz) [reply zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:194]: Will fix this in #948

**Pull request merged**